### PR TITLE
Implement limit

### LINKED ISSUE #676: Support limit and group by
Subtask of #492
- LIMIT syntax: 
LIMIT size/LIMIT skip, size
- GROUP BY syntax
  - syntax
  GROUP BY COUNT(propName), SUM(propName) ……
  - functions
 COUNT、COUNT DISTINCT、 SUM、 AVG、 MIN、MAX、STD
BIT_AND(), BIT_OR(), and BIT_XOR()

### PR DESCRIPTION
**Pull request by laura-ding**
Closes #676

### CODE REVIEW

File: src/graph/LimitExecutor.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    auto inRows = inputs_->getRows();
+    if (inRows.size() > (skip_ + count_)) {
+        rows_.resize(count_);
+        rows_.assign(inRows.begin() + skip_, inRows.begin() + skip_ + count_);
+    } else if (inRows.size() > skip_ && inRows.size() <= (skip_ + count_)) {
+        rows_.resize(inRows.size() - skip_);
+        rows_.assign(inRows.begin() + skip_, inRows.end());
+    } else {
+        rows_.resize(0);
```
Comment (wadeliuyi) [inline src/graph/LimitExecutor.cpp]: it is useless

File: src/parser/test/ParserTest.cpp [hunk 2]
```diff
@@ -1098,4 +1098,25 @@ TEST(Parser, ConfigOperation) {
...
+TEST(Parser, Limit) {
+    {
+        GQLParser parser;
+        std::string query = "GO FROM 1 OVER work | LIMIT 1";
+        auto result = parser.parse(query);
+        ASSERT_TRUE(result.ok()) << result.status();
+    }
+    {
+        GQLParser parser;
+        std::string query = "GO FROM 1 OVER work | LIMIT 1,2";
```
Comment (whitewum) [inline src/parser/test/ParserTest.cpp]: Great. But what do you think about `offset`:
select * from table limit 2,1;  
select * from table limit 2 offset 1;

File: src/parser/test/ParserTest.cpp [hunk 3]
```diff
@@ -1098,4 +1098,25 @@ TEST(Parser, ConfigOperation) {
     }
 }
 
+TEST(Parser, Limit) {
+    {
+        GQLParser parser;
+        std::string query = "GO FROM 1 OVER work | LIMIT 1";
```
Comment (whitewum) [inline src/parser/test/ParserTest.cpp]: since in 99% case, limit works together with order by, can you test them together, it's simple -- the parser test , I mean

File: src/parser/TraverseSentences.h [hunk 4]
```diff
@@ -462,6 +462,28 @@ class FetchEdgesSentence final : public Sentence {
...
+     int64_t skip() {
+        return skip_;
+    }
+
+     int64_t count() {
+        return count_;
+    }
+
+ private:
+    int64_t    skip_{-1};
```
Comment (whitewum) [inline src/parser/TraverseSentences.h]: you alse define skip_ in LimiteExecutor with initial number 0. why not the same.

File: src/graph/LimitExecutor.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+
+Status LimitExecutor::prepare() {
+    skip_ = sentence_->skip();
+    if (skip_ < 0) {
+        return Status::Error("skip `%ld' is illegal", skip_);
+    }
+    count_ = sentence_->count();
+    if (count_ < 0) {
```
Comment (whitewum) [inline src/graph/LimitExecutor.cpp]: i wonder if count_ == 0 has any meaning ? return all or return nothing.  You can add comment in the code to explain.

File: src/graph/test/GroupByLimitTest.cpp [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    void TearDown() override {
+        // ...
+        TraverseTestBase::TearDown();
+    }
+};
+
+TEST_F(GroupByLimitTest, SyntaxError) {
+    {
+        cpp2::ExecutionResponse resp;
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: test : skip < 0 or == 0
test : counter < 0 or == 0
test the combinations above.
They are simple, but worthy corner cases.

File: src/graph/test/GroupByLimitTest.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        auto code = client_->execute(query, resp);
+        ASSERT_EQ(cpp2::ErrorCode::SUCCEEDED, code);
+    }
+}
+
+TEST_F(GroupByLimitTest, LimitTest) {
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Marco Belinelli"];
+        auto *fmt = "GO FROM %ld OVER serve YIELD $$.team.name as name"
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: as -> AS

File: src/graph/test/GroupByLimitTest.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+        auto code = client_->execute(query, resp);
+        ASSERT_EQ(cpp2::ErrorCode::SUCCEEDED, code);
+        std::vector<std::tuple<std::string>> expected = {
+                {"76ers"},
+                {"Bulls"},
+                {"Hawks"},
+                {"Hornets"},
+                {"Kings"},
+        };
+        ASSERT_TRUE(verifyResult(resp, expected, false));
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: let's test a case : the return result number is larger than limit count.

eg. return 5 results but limit 6,1.

File: src/graph/test/GroupByLimitTest.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+                {"Hawks"},
+                {"Hornets"},
+        };
+        ASSERT_TRUE(verifyResult(resp, expected, false));
+    }
+    // test pipe output
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Marco Belinelli"];
+        auto *fmt = "GO FROM %ld OVER like YIELD $$.player.name as name, like._dst as id "
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: as

File: src/graph/test/GroupByLimitTest.cpp [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                    "| ORDER BY $-.name | LIMIT 1 "
+                    "| GO FROM $-.id OVER like YIELD $$.player.name as name "
+                    "| ORDER BY $-.name | LIMIT 2";
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: good.

File: src/graph/test/GroupByLimitTest.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    }
+    // Test less limit
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Danny Green"];
+        auto *fmt = "GO FROM %ld OVER serve YIELD $$.team.name as name | LIMIT 5";
+        auto query = folly::stringPrintf(fmt, player.vid());
+        auto code = client_->execute(query, resp);
+        ASSERT_EQ(cpp2::ErrorCode::SUCCEEDED, code);
+        ASSERT_EQ(3, resp.get_rows()->size());
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: aha, good.
but can you check the detailed result is correct?

File: src/graph/test/GroupByLimitTest.cpp [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+#include "graph/test/TestBase.h"
+#include "graph/test/TraverseTestBase.h"
+#include "meta/test/TestUtils.h"
+
+DECLARE_int32(load_data_interval_secs);
+
+namespace nebula {
+namespace graph {
+
+class GroupByLimitTest : public TraverseTestBase {
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: BTW, why name this file as `GroupBy` `Limit` test?
I think it is not about `GroupBy`

File: src/graph/test/GroupByLimitTest.cpp [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                {"Hornets"},
+                {"Kings"},
+        };
+        ASSERT_TRUE(verifyResult(resp, expected, false));
+    }
+    // Test limit skip,count
+    {
+        cpp2::ExecutionResponse resp;
+        auto &player = players_["Marco Belinelli"];
+        auto *fmt = "GO FROM %ld OVER serve YIELD $$.team.name as name | "
```
Comment (whitewum) [inline src/graph/test/GroupByLimitTest.cpp]: as

File: src/graph/test/GroupByLimitTest.cpp (hunk 12 above)
Comment (laura-ding) [reply src/graph/test/GroupByLimitTest.cpp]: it use the same test file with #749, it will combine if it merge

File: src/graph/LimitExecutor.cpp (hunk 1 above)
Comment (laura-ding) [reply src/graph/LimitExecutor.cpp]: good catch

File: src/graph/LimitExecutor.cpp [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    auto ret = inputs_->getRows();
+    if (!ret.ok()) {
+        DCHECK(onFinish_);
+        onFinish_();
+        return;
+    }
+    auto inRows = std::move(ret).value();
+    if (inRows.size() > static_cast<uint64_t>(skip_ + count_)) {
```
Comment (monadbobo) [inline src/graph/LimitExecutor.cpp]: Perhaps modifying it to the following will make the logic clearer(pseudocode)?
   
``` cpp

   auto size = inRows.size();
    auto real_count = count_ + skip_;
    auto count = 0;
    if (size > skip_) {
        if (size > real_count) {
            count = count_;
        } else {
            count = size;
        }
        rows_.resize(count);
        std::copy_n(inRows.begin() + skip_, rows_.begin(), count);        
    }
```

File: src/graph/LimitExecutor.cpp [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    auto ret = inputs_->getRows();
+    if (!ret.ok()) {
+        DCHECK(onFinish_);
+        onFinish_();
+        return;
+    }
+    auto inRows = std::move(ret).value();
+    if (inRows.size() > static_cast<uint64_t>(offset_ + count_)) {
+        rows_.resize(count_);
+        rows_.assign(inRows.begin() + offset_, inRows.begin() + offset_ + count_);
```
Comment (monadbobo) [inline src/graph/LimitExecutor.cpp]: This should use move_iterator.

File: src/graph/LimitExecutor.cpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    } else if (inRows.size() > static_cast<uint64_t>(offset_) &&
+                   inRows.size() <= static_cast<uint64_t>(offset_ + count_)) {
+        rows_.resize(inRows.size() - offset_);
+        rows_.assign(inRows.begin() + offset_, inRows.end());
```
Comment (monadbobo) [inline src/graph/LimitExecutor.cpp]: the same as above.

**Comment by laura-ding** [comment]
jenkins go
